### PR TITLE
kubelet: Make probe to be on for time drift

### PR DESCRIPTION
**Pull request by pperiyasamy**
The `initialDelaySeconds` holds readiness probes for specified seconds after container has started. To achieve this, kubelet relies on `StartedAt` time of container which never change during its lifecycle. But in case of time drift scenario (example: During maintenance, NVRAM is flashed, looks like it resets the BIOS settings and RTC is also affected), The `StartedAt` time may contain future time which makes probe to be disabled until system reaches `StartedAt` time. Hence this commit handles this scenario and makes probe to be still initiated for those cases.

#### What type of PR is this?

/kind bug

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by sftim** [comment]
Does this PR need a changelog entry?

File: pkg/kubelet/prober/worker.go [hunk 1]
```diff
@@ -265,8 +265,11 @@ func (w *worker) doProbe(ctx context.Context) (keepGoing bool) {
 		return false
 	}
 
-	// Probe disabled for InitialDelaySeconds.
-	if int32(time.Since(c.State.Running.StartedAt.Time).Seconds()) < w.spec.InitialDelaySeconds {
+	// Probe disabled for InitialDelaySeconds only if container's StartedAt time is
+	// earlier than system current time. But probe must be on for time drift case,
+	// This may happen due to nvram flash which resets system BIOS settings and RTC.
+	startedAtInSecs := time.Since(c.State.Running.StartedAt.Time).Seconds()
+	if startedAtInSecs > 0 && int32(startedAtInSecs) < w.spec.InitialDelaySeconds {
```
Comment (bart0sh) [inline pkg/kubelet/prober/worker.go:272]: Would it make sense to wait `InitialDelaySeconds` since `doProbe` called first time if `startedAtInSecs <= 0`.

Comment (pperiyasamy) [reply pkg/kubelet/prober/worker.go:272]: @bart0sh good point, but would that happen (i.e. pod startup followed by immediate time drift event before `InitialDelaySeconds`) ? In our case, it just happened during node maintenance.

**Comment by haircommander** [comment]
out of curiosity: is the the only bug you found in this scenerio? I feel like there would be a lot of little issues with this situation (and I'm not even sure if this is something that the kubelet should support...)

File: pkg/kubelet/prober/worker.go (hunk 1 above)
Comment (bart0sh) [reply pkg/kubelet/prober/worker.go:272]: My point was that if `startedAtInSecs <= 0` probe would run immediately, which is not desirable. It's better to wait for InitialDelaySeconds just to be on the safe side.

Comment (pperiyasamy) [reply pkg/kubelet/prober/worker.go:289]: ok @bart0sh , added code to wait for `InitialDelaySeconds` when time drift happens.

**Comment by pperiyasamy** [comment]
> out of curiosity: is the the only bug you found in this scenerio? I feel like there would be a lot of little issues with this situation (and I'm not even sure if this is something that the kubelet should support...)

@haircommander The readiness probe failure happened on networking pod, so it prevented K8s cluster itself not to come up properly. But after recreation of networking pod, cluster came up just fine.   @brunogomes011 might have more details about this.

/cc @kyrtapz @jcaamano

File: pkg/kubelet/prober/worker.go [hunk 2]
```diff
@@ -70,6 +70,13 @@ type worker struct {
 	// If set, skip probing.
 	onHold bool
 
+	// Set to true when time drift is detected for the pod, will be unset
+	// once container's elapsed start time reaches +ve value.
+	timeDrift bool
+	// timeDriftStartTimeInSec contains container's elapsed start time at
+	// the time of first occurance of time drift.
```
Comment (bart0sh) [inline pkg/kubelet/prober/worker.go:77]:
```suggestion
	// the time of first occurrence of time drift.
```

**Comment by bart0sh** [comment]
@pperiyasamy I agree with @haircommander here. Keeping machine system clock  is a basic administration task. I'm not sure Kubelet should work around incorrectly set system clock.

**Comment by sftim** [comment]
> check this

One way to check: make a request to the API server, look at the `Date:` HTTP header, see if it's within _n_ seconds of the local UTC time.

**Pull request closed without merging**